File: src/page/info/filter/UnitEditBox.java
package page.info.filter;

import static common.util.Interpret.ABIIND;
import static common.util.Interpret.IMUSFT;
import static common.util.Interpret.SABIS;
import static common.util.Interpret.SPROC;
import static common.util.Interpret.TRAIT;

import java.util.Vector;

import javax.swing.JList;
import javax.swing.JScrollPane;
import javax.swing.ListSelectionModel;
import javax.swing.event.ListSelectionEvent;
import javax.swing.event.ListSelectionListener;

import page.Page;

public class UnitEditBox extends Page {

	private static final long serialVersionUID = 1L;

	private final boolean editable;

	private final Vector<String> vt = new Vector<>();
	private final Vector<String> va = new Vector<>();
	private final AttList trait = new AttList(3, 0);
	private final AttList abis = new AttList(0, 1);
	private final JScrollPane jt = new JScrollPane(trait);
	private final JScrollPane jab = new JScrollPane(abis);

	private boolean changing = false;

	public UnitEditBox(Page p, boolean edit) {
		super(p);
		editable = edit;
		ini();
	}

	public void setData(int[] vals) {
		changing = true;
		trait.clearSelection();
		abis.clearSelection();
		for (int i = 0; i < 9; i++)
			if (((vals[0] >> i) & 1) > 0)
				trait.addSelectionInterval(i, i);
		for (int i = 0; i < SABIS.length; i++) {
			if (((vals[1] >> i) & 1) > 0)
				abis.addSelectionInterval(i, i);
		}
		int lev = SABIS.length;
		for (int i = 0; i < ABIIND.length; i++) {
			int ind = ABIIND[i];
			if (((vals[2] >> (ind - 100 - IMUSFT)) & 1) > 0)
				abis.addSelectionInterval(lev + i, lev + i);
		}
		changing = false;
	}

	@Override
	protected void resized(int x, int y) {
		set(jt, x, y, 0, 0, 200, 400);
		set(jab, x, y, 0, 450, 200, 750);
	}

	private void confirm() {
		int[] ans = new int[3];
		for (int i = 0; i < 9; i++)
			if (trait.isSelectedIndex(i))
				ans[0] |= 1 << i;
		int lev = SABIS.length;
		for (int i = 0; i < lev; i++)
			if (abis.isSelectedIndex(i))
				ans[1] |= 1 << i;
		for (int i = 0; i < ABIIND.length; i++)
			if (abis.isSelectedIndex(lev + i))
				ans[2] |= 1 << (ABIIND[i] - 100 - IMUSFT);

		getFront().callBack(ans);
	}

	private void ini() {
		for (int i = 0; i < 9; i++)
			vt.add(TRAIT[i]);
		for (int i = 0; i < SABIS.length; i++)
			va.add(SABIS[i]);
		for (int i = 0; i < ABIIND.length; i++)
			va.add(SPROC[ABIIND[i] - 100]);
		trait.setListData(vt);
		abis.setListData(va);
		int m = ListSelectionModel.MULTIPLE_INTERVAL_SELECTION;
		trait.setSelectionMode(m);
		abis.setSelectionMode(m);
		set(trait);
		set(abis);
		add(jt);
		add(jab);
		trait.setEnabled(editable);
		abis.setEnabled(editable);
	}

	private void set(JList<?> jl) {

		jl.addListSelectionListener(new ListSelectionListener() {

			@Override
			public void valueChanged(ListSelectionEvent arg0) {
				if (!changing && !jl.getValueIsAdjusting())
					confirm();
			}

		});
	}

}

File: src/page/info/filter/UnitEditBox.kt
package page.info.filter

import page.Page
import utilpc.Interpret
import java.util.*
import javax.swing.JList
import javax.swing.JScrollPane
import javax.swing.ListSelectionModel
import javax.swing.event.ListSelectionEvent
import javax.swing.event.ListSelectionListener

class UnitEditBox(p: Page?, private val editable: Boolean) : Page(p) {
    private val vt = Vector<String>()
    private val va = Vector<String>()
    private val trait = AttList(3, 0)
    private val abis = AttList(0, 1)
    private val jt: JScrollPane = JScrollPane(trait)
    private val jab: JScrollPane = JScrollPane(abis)
    private var changing = false
    fun setData(vals: IntArray) {
        changing = true
        trait.clearSelection()
        abis.clearSelection()
        for (i in 0..8) if (vals[0] shr i and 1 > 0) trait.addSelectionInterval(i, i)
        for (i in Interpret.SABIS.indices) {
            if (vals[1] shr i and 1 > 0) abis.addSelectionInterval(i, i)
        }
        val lev: Int = Interpret.SABIS.size
        for (i in Interpret.ABIIND.indices) {
            val ind: Int = Interpret.ABIIND.get(i)
            if (vals[2] shr ind - 100 - Interpret.IMUSFT and 1 > 0) abis.addSelectionInterval(lev + i, lev + i)
        }
        changing = false
    }

    override fun resized(x: Int, y: Int) {
        Page.Companion.set(jt, x, y, 0, 0, 200, 400)
        Page.Companion.set(jab, x, y, 0, 450, 200, 750)
    }

    private fun confirm() {
        val ans = IntArray(3)
        for (i in 0..8) if (trait.isSelectedIndex(i)) ans[0] = ans[0] or (1 shl i)
        val lev: Int = Interpret.SABIS.size
        for (i in 0 until lev) if (abis.isSelectedIndex(i)) ans[1] = ans[1] or (1 shl i)
        for (i in Interpret.ABIIND.indices) if (abis.isSelectedIndex(lev + i)) ans[2] = ans[2] or (1 shl Interpret.ABIIND.get(i) - 100 - Interpret.IMUSFT)
        front.callBack(ans)
    }

    private fun ini() {
        for (i in 0..8) vt.add(Interpret.TRAIT.get(i))
        for (i in Interpret.SABIS.indices) va.add(Interpret.SABIS.get(i))
        for (i in Interpret.ABIIND.indices) va.add(Interpret.SPROC.get(Interpret.ABIIND.get(i) - 100))
        trait.setListData(vt)
        abis.setListData(va)
        val m: Int = ListSelectionModel.MULTIPLE_INTERVAL_SELECTION
        trait.selectionMode = m
        abis.selectionMode = m
        set(trait)
        set(abis)
        add(jt)
        add(jab)
        trait.isEnabled = editable
        abis.isEnabled = editable
    }

    private fun set(jl: JList<*>) {
        jl.addListSelectionListener(object : ListSelectionListener {
            override fun valueChanged(arg0: ListSelectionEvent?) {
                if (!changing && !jl.valueIsAdjusting) confirm()
            }
        })
    }

    companion object {
        private const val serialVersionUID = 1L
    }

    init {
        ini()
    }
}
